### PR TITLE
Sequences

### PR DESCRIPTION
**Pull request by exyi**
Adds two relations

* **sequences** - contains the metadata about sequences
    - supports insert, update and delete. Upsert is problematic, since sequences don't return the nice 409 error when ids already exist.
    - does not do partitioned read nor filter pushdown (no support in API, no support in Scala SDK).
* **sequenceRows** - the sequence rows
    - requires `sequenceId` or `sequenceExternalId` option that specifies the sequence that should be used
    - when created, it fetches the sequence metadata and the Spark schema will correspond to that
    - supports upsert, delete and read
    - does filter and projection pushdown, does not do any partitioning

### CODE REVIEW

File: src/main/scala/cognite/spark/v1/DefaultSource.scala [hunk 1]
```diff
@@ -102,6 +117,8 @@ class DefaultSource
           inferSchema,
           inferSchemaLimit,
           collectSchemaInferenceMetrics)(sqlContext)
+      case "sequenceRows" =>
```
Comment (wjoel) [inline src/main/scala/cognite/spark/v1/DefaultSource.scala:120]: We use lowercase for those, `sequencerows` (just like `3dmodelrevisionnodes` etc.)

File: src/main/scala/cognite/spark/v1/DefaultSource.scala [hunk 2]
```diff
@@ -153,6 +172,10 @@ class DefaultSource
           new NumericDataPointsRelationV1(config)(sqlContext)
         case "stringdatapoints" =>
           new StringDataPointsRelationV1(config)(sqlContext)
+        case "sequences" =>
+          new SequencesRelation(config)(sqlContext)
+        case "sequenceRows" =>
```
Comment (wjoel) [inline src/main/scala/cognite/spark/v1/DefaultSource.scala:177]: We use lowercase for those, `sequencerows` (just like `stringdatapoints` etc.)

File: src/main/scala/cognite/spark/v1/SequenceRowsRelation.scala [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+
+  def fromRow(schema: StructType): (Array[String], Row => SequenceRow) = {
+    val rowNumberIndex = schema.fieldNames.indexOf("rowNumber")
+    if (rowNumberIndex < 0) {
+      throw new Exception("Can't upsert sequence rows, column `rowNumber` is missing.")
+    }
+
+    val columns = schema.fields.zipWithIndex.filter(_._1.name != "rowNumber").map {
+      case (field, index) =>
+        val seqColumn = sequenceInfo.columns
```
Comment (wjoel) [inline src/main/scala/cognite/spark/v1/SequenceRowsRelation.scala:84]: Store `sequenceInfo.columns` as a `Set` and use `contains`.

File: src/main/scala/cognite/spark/v1/SequenceRowsRelation.scala [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+          .find(_.externalId == field.name)
+          .getOrElse(throw new Exception(
+            s"Can't insert column `${field.name}` into sequence $sequenceId, the column does not exist in the sequence definition"))
+        def jsonFromDouble(num: Double): Json =
+          Json.fromDouble(num).getOrElse(throw new Exception(s"Numeric value $num"))
+        val tryGetValue: PartialFunction[Any, Json] = seqColumn.valueType match {
```
Comment (wjoel) [inline src/main/scala/cognite/spark/v1/SequenceRowsRelation.scala:90]: Move those to `private def`s (or `private val tryGetValue` if necessary)

File: src/main/scala/cognite/spark/v1/SequenceRowsRelation.scala [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+    val rowNumberIndex = columns.indexOf("rowNumber")
+
+    SdkV1Rdd[ProjectedSequenceRow, Long](
+      sqlContext.sparkContext,
+      configWithLimit,
+      (item: ProjectedSequenceRow) => {
+        if (config.collectMetrics) {
+          itemsRead.inc()
+        }
+        Row.fromSeq(if (rowNumberIndex < 0) {
```
Comment (wjoel) [inline src/main/scala/cognite/spark/v1/SequenceRowsRelation.scala:166]: When/how does this happen?

File: src/main/scala/cognite/spark/v1/SequenceRowsRelation.scala [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+          val (beforeRowNumber, afterRowNumber) = item.values.splitAt(rowNumberIndex)
+          beforeRowNumber ++ Array(item.rowNumber) ++ afterRowNumber
+        })
+      },
+      (r: ProjectedSequenceRow) => r.rowNumber,
+      getStreams(filters, columns.filter(_ != "rowNumber"))
+    )
+  }
+
+//  override def buildScan(): RDD[Row] = buildScan(Array.empty, Array.empty)
```
Comment (wjoel) [inline src/main/scala/cognite/spark/v1/SequenceRowsRelation.scala:178]: Delete

File: src/main/scala/cognite/spark/v1/SequenceRowsRelation.scala [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+
+  override def buildScan(requiredColumns: Array[String], filters: Array[Filter]): RDD[Row] = {
+
```
Comment (wjoel) [inline src/main/scala/cognite/spark/v1/SequenceRowsRelation.scala:181]: Delete empty row

File: src/main/scala/cognite/spark/v1/SequenceRowsRelation.scala [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+        .reduceOption(filterIntersection)
+        .getOrElse(Seq(SequenceRowFilter()))
+
+    val rdd =
+      readRows(
+        config.limitPerPartition,
+        filterObjects,
+        requiredColumns
+      )
+    rdd
```
Comment (wjoel) [inline src/main/scala/cognite/spark/v1/SequenceRowsRelation.scala:194]: Just return `readRows(...)` as you don't use the val

File: src/main/scala/cognite/spark/v1/SequenceRowsRelation.scala [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+
+  private def toSegments(f: Vector[SequenceRowFilter]) = {
+    val (minusInfCount, plusInfCount, borders) = toBorders(f.toVector)
+    // count number of overlapping intervals in each segment
+    val segmentValues = borders.scanLeft[Int, Vector[Int]](minusInfCount)((v, b) => {
+      if (b.start) {
+        v + 1
+      } else {
+        v - 1
+      }
```
Comment (wjoel) [inline src/main/scala/cognite/spark/v1/SequenceRowsRelation.scala:246]: Please add comments about why we do this.

File: src/main/scala/cognite/spark/v1/SequenceRowsRelation.scala [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+    })
+    val borderLabels = Vector(None) ++ borders.map(b => Some(b.value)) ++ Vector(None)
+    val segmentLabels =
+      borderLabels
+        .zip(borderLabels.drop(1))
```
Comment (wjoel) [inline src/main/scala/cognite/spark/v1/SequenceRowsRelation.scala:251]: Add a comment explaining `drop(1)`

File: src/main/scala/cognite/spark/v1/SequencesRelation.scala [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+    extends SdkV1Relation[SequenceReadSchema, Long](config, "sequences")
+    with InsertableRelation
+    with WritableRelation {
+  import CdpConnector._
+
+  override def getStreams(filters: Array[Filter])(
+      client: GenericClient[IO, Nothing],
+      limit: Option[Int],
+      numPartitions: Int): Seq[Stream[IO, SequenceReadSchema]] =
+    // TODO: partitions and filters!!!
```
Comment (wjoel) [inline src/main/scala/cognite/spark/v1/SequencesRelation.scala:28]: I don't think they support partitions?

File: src/main/scala/cognite/spark/v1/SequencesRelation.scala [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+      doUpsert = true)
+  }
+
+  override def schema: StructType = structType[SequenceReadSchema]
+
+  override def toRow(a: SequenceReadSchema): Row = asRow(a)
+
+  override def uniqueId(a: SequenceReadSchema): Long = a.id
+}
+object SequenceRelation extends UpsertSchema {
```
Comment (wjoel) [inline src/main/scala/cognite/spark/v1/SequencesRelation.scala:110]: Add empty line before `object` declaration.

File: src/test/scala/cognite/spark/v1/SequenceRowsRelationTest.scala [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+      .parallelize(1 to 100)
+      .toDF()
+      .createOrReplaceTempView("numbers")
+
+    insertRows(
+      sequenceId,
+      spark
+        .sql("select value as rowNumber, 'abc' as str1, 1.1 as num2, value * 6 as num1 from numbers"))
+
+    val allColumns = retryWhile[Array[Row]](
```
Comment (wjoel) [inline src/test/scala/cognite/spark/v1/SequenceRowsRelationTest.scala:66]: Sequences are *not* eventually consistent, so the data should be available immediately, but I guess the retry doesn't hurt.

File: src/test/scala/cognite/spark/v1/SequenceRowsRelationTest.scala [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+    differentOrderProjection(0).get(1) shouldBe 1.1
+    differentOrderProjection(0).get(0) shouldBe 6L
+
+    val oneColumn = retryWhile[Array[Row]](
+      spark.sql(s"select num2 from sequenceRows_a order by rowNumber").collect,
+      _.length != 100
+    )
+    oneColumn.map(_.get(0)) shouldBe Seq.fill(100)(1.1)
+  }
+  it should "insert NULL values" in withSequences(Seq(sequenceA)) { case Seq(sequenceId) =>
```
Comment (wjoel) [inline src/test/scala/cognite/spark/v1/SequenceRowsRelationTest.scala:97]: Empty line to separate `it should` clauses.

File: src/test/scala/cognite/spark/v1/SequenceRowsRelationTest.scala [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+      spark.sql(s"select * from sequenceRows_b order by rowNumber").collect,
+      _.length != testSize
+    )
+    allColumns(0).schema.fieldNames shouldBe Array("rowNumber", "num1")
+    allColumns.map(_.getAs[Long]("num1")) shouldBe (1 to testSize).map(_ * 6)
+  }
+
+  // ----------
+
+  def withSequences(sequences: Seq[SequenceUpdateSchema])(testCode: Seq[String] => Unit): Unit = {
```
Comment (wjoel) [inline src/test/scala/cognite/spark/v1/SequenceRowsRelationTest.scala:195]: Nice!

File: src/main/scala/cognite/spark/v1/SequenceRowsRelation.scala [hunk 16]
```diff
@@ -0,0 +1,318 @@
...
+                .badRowError(row, seqColumn.externalId, seqColumn.valueType, "")
+          )
+      }
+      SequenceRow(rowNumber, columnValues)
+    }
+
+    (columns.map(_._2.externalId), parseRow)
+  }
+
+  def delete(rows: Seq[Row]): IO[Unit] = ???
```
Comment (wjoel) [inline src/main/scala/cognite/spark/v1/SequenceRowsRelation.scala]: Woops? :) Can we support deletes in an easy way? Probably not yet (would be similar to data points deletes, I suppose), in which case it's fine to throw an exception about it not being supported.

File: src/test/scala/cognite/spark/v1/SequenceRowsRelationTest.scala [hunk 17]
```diff
@@ -0,0 +1,243 @@
...
+      .format("cognite.spark.v1")
+      .option("apiKey", writeApiKey)
+      .option("type", "sequenceRows")
+      .option("sequenceExternalId", externalId)
+      .load()
+
+  def insertRows(seqId: String, df: DataFrame) =
+    df.write
+      .format("cognite.spark.v1")
+      .option("type", "sequenceRows")
```
Comment (wjoel) [inline src/test/scala/cognite/spark/v1/SequenceRowsRelationTest.scala:219]: Use lowercase `sequencerows`

File: src/test/scala/cognite/spark/v1/SequenceRowsRelationTest.scala [hunk 18]
```diff
@@ -0,0 +1,243 @@
...
+    } finally {
+      cleanupSequence(key, sequences.map(_.externalId.get) :_*)
+    }
+  }
+
+  def createRowsRelation(externalId: String) =
+    spark.read
+      .format("cognite.spark.v1")
+      .option("apiKey", writeApiKey)
+      .option("type", "sequenceRows")
```
Comment (wjoel) [inline src/test/scala/cognite/spark/v1/SequenceRowsRelationTest.scala:212]: Use lowercase `sequencerows`

File: src/main/scala/cognite/spark/v1/SequenceRowsRelation.scala [hunk 19]
```diff
@@ -0,0 +1,318 @@
...
+  def insert(rows: Seq[Row]): IO[Unit] =
+    throw new Exception("Insert not supported for sequenceRows. Use upsert instead.")
+  def update(rows: Seq[Row]): IO[Unit] =
+    throw new Exception("Update not supported for sequenceRows. Use upsert instead.")
```
Comment (wjoel) [inline src/main/scala/cognite/spark/v1/SequenceRowsRelation.scala:141]: Use lowercase `sequencerows`

File: src/test/scala/cognite/spark/v1/SequenceRowsRelationTest.scala [hunk 20]
```diff
@@ -0,0 +1,243 @@
...
+    createRowsRelation(sequenceId).createOrReplaceTempView("sequenceRows_b")
+
+    val allColumns = retryWhile[Array[Row]](
+      spark.sql(s"select * from sequenceRows_b order by rowNumber").collect,
+      _.length != testSize
+    )
+    allColumns(0).schema.fieldNames shouldBe Array("rowNumber", "num1")
+    allColumns.map(_.getAs[Long]("num1")) shouldBe (1 to testSize).map(_ * 6)
+  }
+
```
Comment (wjoel) [inline src/test/scala/cognite/spark/v1/SequenceRowsRelationTest.scala:162]: Add tests covering the pushdown filters.

File: src/main/scala/cognite/spark/v1/SequenceRowsRelation.scala (hunk 3 above)
Comment (exyi) [reply src/main/scala/cognite/spark/v1/SequenceRowsRelation.scala:84]: Turns out I already had the `columnTypes` map :)

File: src/main/scala/cognite/spark/v1/SequencesRelation.scala (hunk 11 above)
Comment (exyi) [reply src/main/scala/cognite/spark/v1/SequencesRelation.scala:28]: They don't, unfortunatelly :/ Hopefully there is not gonna be too many sequences in a single project

File: src/test/scala/cognite/spark/v1/SequenceRowsRelationTest.scala (hunk 13 above)
Comment (exyi) [reply src/test/scala/cognite/spark/v1/SequenceRowsRelationTest.scala:66]: Are you sure? I have just tried, and it seems that it just does not return any rows right after the insert :/ I'll just keep the retries here, but would be nice to remove them.

File: src/main/scala/cognite/spark/v1/SequenceRowsRelation.scala (hunk 16 above)
Comment (exyi) [reply src/main/scala/cognite/spark/v1/SequenceRowsRelation.scala]: Sequence rows are deleted by the `rowNumber`, so I have just added that.

File: src/main/scala/cognite/spark/v1/DefaultSource.scala [hunk 21]
```diff
@@ -67,6 +67,21 @@ class DefaultSource
   override def createRelation(sqlContext: SQLContext, parameters: Map[String, String]): BaseRelation =
     createRelation(sqlContext, parameters, null) // scalastyle:off null
 
+  private def createSequenceRows(parameters: Map[String, String], config: RelationConfig, sqlContext: SQLContext) = {
+    val sequenceId =
+      parameters.get("sequenceId").map(id => CogniteInternalId(id.toInt))
+        .orElse(
+          parameters.get("sequenceExternalId").map(CogniteExternalId)
```
Comment (wjoel) [inline src/main/scala/cognite/spark/v1/DefaultSource.scala:74]: Isn't `sequence` a bit redundant here, if the user has doen `.option("type", "sequencerows").option("sequenceid", "1828")`?

File: src/main/scala/cognite/spark/v1/SequenceRowsRelation.scala (hunk 10 above)
Comment (wjoel) [reply src/main/scala/cognite/spark/v1/SequenceRowsRelation.scala:260]: Still doesn't say why :P But ok, something with the zipping.

File: src/test/scala/cognite/spark/v1/SequenceRowsRelationTest.scala (hunk 20 above)
Comment (exyi) [reply src/test/scala/cognite/spark/v1/SequenceRowsRelationTest.scala:182]: This was a very good point :) Added the test and fixed like 3 bugs in there :facepalm:

File: src/main/scala/cognite/spark/v1/SequenceRowsRelation.scala [hunk 22]
```diff
@@ -0,0 +1,306 @@
...
+  def delete(rows: Seq[Row]): IO[Unit] = {
+    val deletes = rows.map(r => SparkSchemaHelper.fromRow[SequenceRowDeleteSchema](r))
+    client.sequenceRows.deleteById(sequenceInfo.id, deletes.map(_.rowNumber))
+      .flatTap(_ => incMetrics(itemsDeleted, rows.length))
+  }
+  def insert(rows: Seq[Row]): IO[Unit] =
+    throw new Exception("Insert not supported for sequencerows. Use upsert instead.")
+  def update(rows: Seq[Row]): IO[Unit] =
+    throw new Exception("Update not supported for sequencerows. Use upsert instead.")
+  def upsert(rows: Seq[Row]): IO[Unit] = {
```
Comment (wjoel) [inline src/main/scala/cognite/spark/v1/SequenceRowsRelation.scala:151]: Empty lines to separate defs.

File: src/test/scala/cognite/spark/v1/SequenceRowsRelationTest.scala (hunk 13 above)
Comment (wjoel) [reply src/test/scala/cognite/spark/v1/SequenceRowsRelationTest.scala:68]: Ah, it's like data points... the storage is not eventually consistent, but there's a pubsub queue in the way, so for the user it is indeed eventually consistent. :/

File: src/main/scala/cognite/spark/v1/DefaultSource.scala (hunk 21 above)
Comment (exyi) [reply src/main/scala/cognite/spark/v1/DefaultSource.scala:74]: Hmm, maybe :thinking: I still find saying just `id` might not be very clear. Should I change it, or?

Comment (wjoel) [reply src/main/scala/cognite/spark/v1/DefaultSource.scala:74]: I think so tbh. I think it is just as clear as "sequenceid", and doubt there will be any confusion.

Comment (exyi) [reply src/main/scala/cognite/spark/v1/DefaultSource.scala]: Ok, I have changed that (and added some docs)

File: README.md [hunk 23]
```diff
@@ -384,6 +384,32 @@ The schemas mirror the CDF API as closely as possible.
 | `metadata`           | `map(string, string)` | Yes       |
 | `dataSetId`          | `long`                | Yes       |
 
+### Sequences
```
Comment (wjoel) [inline README.md:387]: Link those in the table of contents as well, similar to `- [Asset Hierarchy](#asset-hierarchy)`

File: README.md [hunk 24]
```diff
@@ -384,6 +384,32 @@ The schemas mirror the CDF API as closely as possible.
...
+
+| Column name          | Type                  |  Nullable |
+| ---------------------| ----------------------| --------- |
+| `externalId`         | `string`              | No        |
+| `name`   | `string`              | Yes        |
+| `description`             | `string`              | Yes       |
+| `valueType`               | `string`              | No        |
+| `metadata`           | `map(string, string)` | Yes       |
+| `dataSetId`          | `long`                | Yes       |
+| `columns`        | `array(SequenceColumn)`              | No       |
```
Comment (wjoel) [inline README.md:408]: They can be nested?

File: README.md [hunk 25]
```diff
@@ -384,6 +384,32 @@ The schemas mirror the CDF API as closely as possible.
...
+| ---------------------| ----------------------| --------- |
+| `externalId`         | `string`              | Yes        |
+| `name`   | `string`              | Yes        |
+| `description`             | `string`              | Yes       |
+| `assetId`               | `long`              | Yes        |
+| `metadata`           | `map(string, string)` | Yes       |
+| `dataSetId`          | `long`                | Yes       |
+| `columns`        | `array(SequenceColumn)`              | No       |
+
+The `columns` field should be an array of `SequenceColumn`s, which are rows with the following fields:
```
Comment (wjoel) [inline README.md:398]: I don't know how to create that after reading this... what would that look like in SQL? Could we provide an example?

I kind of wonder if it wouldn't be a lot more user friendly to just use an array of maps to describe the columns?

File: src/test/scala/cognite/spark/v1/SequencesRelationTest.scala [hunk 26]
```diff
@@ -0,0 +1,133 @@
...
+    .option("type", "sequences")
+    .load()
+  sequencesSourceDf.createOrReplaceTempView("sequences")
+
+  it should "create and read sequence" in {
+    val key = shortRandomString()
+    val sequence = SequenceUpdateSchema(
+      externalId = Some("a"),
+      name = Some("a"),
+      columns = Some(Seq(
```
Comment (wjoel) [inline src/test/scala/cognite/spark/v1/SequencesRelationTest.scala:28]: How to do this with `spark.sql`?

File: src/test/scala/cognite/spark/v1/SequencesRelationTest.scala [hunk 27]
```diff
@@ -0,0 +1,133 @@
...
+
+    cleanupSequence(key, "a")
+  }
+
+  def ingest(
+              key: String,
+              tree: Seq[SequenceUpdateSchema],
+              metricsPrefix: Option[String] = None,
+              conflictMode: String = "abort"
+  ): Unit = {
```
Comment (wjoel) [inline src/test/scala/cognite/spark/v1/SequencesRelationTest.scala:99]: Bad indentation.

File: src/test/scala/cognite/spark/v1/SequencesRelationTest.scala (hunk 26 above)
Comment (exyi) [reply src/test/scala/cognite/spark/v1/SequencesRelationTest.scala:28]: This is a good point. Turns out it's quite simple, I have added a test and example into the REAMDE.md - you just create `array(named_struct("externalId", "x", ...))`

File: README.md (hunk 25 above)
Comment (exyi) [reply README.md:398]: I have added an example, not sure if I can link it from here.

**Pull request merged**